### PR TITLE
DEPR: Deprecate tshift and integrate it to shift

### LINKED ISSUE #11631: CLN: tshift & shift could be consolidated
Particularly now that `shift` only works on datetime-like indexes (https://github.com/pydata/pandas/pull/11211)

They look almost the same, although not exactly:

``` python
In [11]: df=pd.DataFrame(pd.np.random.rand(5,2), index=pd.date_range(periods=5, start='2000'))

In [12]: df
Out[12]: 
                   0         1
2000-01-01  0.640148  0.781291
2000-01-02  0.261649  0.652372
2000-01-03  0.642422  0.734348
2000-01-04  0.582657  0.601868
2000-01-05  0.848645  0.078437

In [13]: df.shift()
Out[13]: 
                   0         1
2000-01-01       NaN       NaN
2000-01-02  0.640148  0.781291
2000-01-03  0.261649  0.652372
2000-01-04  0.642422  0.734348
2000-01-05  0.582657  0.601868

In [14]: df.tshift()
Out[14]: 
                   0         1
2000-01-02  0.640148  0.781291
2000-01-03  0.261649  0.652372
2000-01-04  0.642422  0.734348
2000-01-05  0.582657  0.601868
2000-01-06  0.848645  0.078437

In [15]: df.shift(freq='D')
Out[15]: 
                   0         1
2000-01-02  0.640148  0.781291
2000-01-03  0.261649  0.652372
2000-01-04  0.642422  0.734348
2000-01-05  0.582657  0.601868
2000-01-06  0.848645  0.078437

In [16]: df.tshift(freq='D')
Out[16]: 
                   0         1
2000-01-02  0.640148  0.781291
2000-01-03  0.261649  0.652372
2000-01-04  0.642422  0.734348
2000-01-05  0.582657  0.601868
2000-01-06  0.848645  0.078437

```

### PR DESCRIPTION
**Pull request by fujiaxiang**
- [x] closes #11631
- xref #34452
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
thanks. would not be averse to actually re-writing tshift in terms of shift; I *think* this would be straightforward. Also could merge this and then followon to do that.

**Comment by jreback** [comment]
> > would not be averse to actually re-writing tshift in terms of shift; I think this would be straightforward
> 
> Agree it's rather straightforward, but just to clarify a few things:
> 
> 1. we are still going to deprecate `tshift` after the re-write right? In other words, you don't mean to re-write and **keep** `tshift` right?

correct, still deprecate

> 2. do you think it's better to move the current `_tshift` inside `shift`? i.e. we won't have a separate method `_tshift`, but unpack what's in it and put inside `shift`

yes if possible (ideally simplified as much as possible).

**Comment by fujiaxiang** [comment]
Have unpacked previous `_tshift` and moved into `shift`, also simplified the code a little bit. The deprecating `tshift` now calls `shift`. @jreback can you review?

File: pandas/tests/frame/methods/test_shift.py [hunk 1]
```diff
@@ -186,10 +189,60 @@ def test_tshift(self, datetime_frame):
...
+        ps = tm.makePeriodFrame()
+        shifted = ps.shift(1, freq="infer")
+        unshifted = shifted.shift(-1, freq="infer")
+
+        tm.assert_frame_equal(unshifted, ps)
+
+        shifted2 = ps.shift(freq="B")
+        tm.assert_frame_equal(shifted, shifted2)
+
+        shifted3 = ps.shift(freq=offsets.BDay())
```
Comment (jreback) [inline pandas/tests/frame/methods/test_shift.py:212]: can you put the failures in a separate test (e.g. things that assert raises, name it like test_shift_with_freq_errors

also not averse to separating PI and DTI into 2 tests. if possibly to parameterize would be great but if not ok too.

File: pandas/tests/series/methods/test_shift.py [hunk 2]
```diff
@@ -220,10 +223,58 @@ def test_tshift(self, datetime_series):
...
+        msg = "Freq was not set in the index hence cannot be inferred"
         with pytest.raises(ValueError, match=msg):
             no_freq.tshift()
 
+    def test_tshift_deprecated(self, datetime_series):
+        # GH#11631
+        with tm.assert_produces_warning(FutureWarning):
+            datetime_series.tshift()
+
+    def test_shift_with_freq(self, datetime_series):
```
Comment (jreback) [inline pandas/tests/series/methods/test_shift.py:235]: same comment as above, e.g. to split this up a bit

**Comment by jreback** [comment]
@fujiaxiang can you rebase and some small comments, ping on green.

File: pandas/tests/frame/methods/test_shift.py (hunk 1 above)
Comment (fujiaxiang) [reply pandas/tests/frame/methods/test_shift.py:212]: done

File: pandas/tests/series/methods/test_shift.py (hunk 2 above)
Comment (fujiaxiang) [reply pandas/tests/series/methods/test_shift.py]: done

**Comment by fujiaxiang** [comment]
@jreback have split up the tests into PI and DTI and also separated the failure tests
I tried to parametrize the test cases into PI and DTI but I realize they only have a very small proportion of code in common, so I decided to separate them into different tests

**Comment by jreback** [comment]
> @jreback have split up the tests into PI and DTI and also separated the failure tests
> I tried to parametrize the test cases into PI and DTI but I realize they only have a very small proportion of code in common, so I decided to separate them into different tests

kk sounds good. ping on green (note that there maybe be a failure on master that is unrelated)

**Comment by jreback** [comment]
thanks @fujiaxiang very nice

**Pull request merged**